src/sources/http.rs

```diff
@@ -0,0 +1,344 @@
+use crate::{
+    event::{self, flatten::flatten, Event},
+    sources::util::{ErrorMessage, HttpSource},
+    topology::config::{DataType, GlobalOptions, SourceConfig},
+};
+use bytes::Buf;
+use chrono::Utc;
+use futures::sync::mpsc;
+use serde::{Deserialize, Serialize};
+use serde_json::Value as JsonValue;
+use std::{
+    io::{BufRead, BufReader},
+    net::SocketAddr,
+};
+use warp::http::{HeaderMap, StatusCode};
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct SimpleHttpConfig {
+    address: SocketAddr,
+    #[serde(default)]
+    encoding: Encoding,
+    #[serde(default)]
+    headers: Vec<String>,
+}
+
+#[derive(Clone)]
+struct SimpleHttpSource {
+    encoding: Encoding,
+    headers: Vec<String>,
+}
+
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Derivative)]
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+pub enum Encoding {
+    #[derivative(Default)]
+    Text,
+    Ndjson,
+    Json,
+}
+
+impl HttpSource for SimpleHttpSource {
+    fn build_event(
+        &self,
+        body: impl Buf,
+        header_map: HeaderMap,
+    ) -> Result<Vec<Event>, ErrorMessage> {
+        decode_body(body, self.encoding.clone())
+            .map(|events| add_headers(events, &self.headers, header_map))
+    }
+}
+
+#[typetag::serde(name = "http")]
+impl SourceConfig for SimpleHttpConfig {
+    fn build(
+        &self,
+        _: &str,
+        _: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        let source = SimpleHttpSource {
+            encoding: self.encoding.clone(),
+            headers: self.headers.clone(),
+        };
+        source.run(self.address, "".to_string(), out)
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "http"
+    }
+}
+
+fn add_headers(
+    mut events: Vec<Event>,
+    headers_config: &Vec<String>,
+    headers: HeaderMap,
+) -> Vec<Event> {
+    for header_name in headers_config {
+        let mut value: &[u8] = &[];
```

Comment:
could avoid `mut` with:

```rust
        let value = if headers.contains_key(header_name) {
            headers[header_name].as_bytes()
        } else {
            &[]
        };
```

`headers_config` could take a slice or Vec if it's type is `&[String]`

Reply:
Or even better probably `get().or_default()`